### PR TITLE
Fix several bugs related to unbacked SymInt codegen in inductor

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118411
* #118125
* __->__ #117862

Let me tell you, this was a *journey.*

* When we repropagate through FX interpreter in AOTAutograd, this will reallocate unbacked SymInts. We can eliminate all of these fresh allocations by appropriately asserting equalities on them setting up replacements. See also https://github.com/pytorch/pytorch/issues/111950
* The `inner_fn` of Loops can contain references to unbacked SymInts. We must collect them to prevent DCE.
* Export naughtily accessed `_expr` when it should have accessed `expr` on SymNode. Fixed two sites of this.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by flaviotruzzi** [comment]
> I wonder if we are generating runtime asserts like i0 == i1 for the duplicative unbacked symbols, that would be bad 🤔

I don't think I saw something like that in our logs and/or debugging. They should be the same but inferred not explicitly.

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 1]
```diff
@@ -587,6 +589,20 @@ def metadata_fn(*primals):
...
+class PropagateUnbackedSymInts(torch.fx.Interpreter):
+    def run_node(self, n: torch.fx.Node):
+        import sympy
+        from torch.fx.experimental.symbolic_shapes import free_unbacked_symbols
+
+        result = super().run_node(n)
+        # TODO: handle Tensor returns
+        if 'example_value' in n.meta:
+            if isinstance(result, torch.SymInt):
+                torch._check(result == n.meta['example_value'])
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:601]: don't we only have to do this check if result has any unbacked symbols in it? (or is it harmless to do this check when it's backed, since presumably it would no-op)

Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:601]: Yeah, is harmless if it's backed

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1312,6 +1312,7 @@ def rename_indexing(self, index) -> sympy.Expr:
             x: self.args.size(x)
             for x in sorted_symbols
             if x.name.startswith("s")
+            or x.name.startswith("u")
             or x.name.startswith("ps")
```
Comment (Skylion007) [inline torch/_inductor/codegen/common.py:1316]:
```suggestion
            if x.name.startswith(("s", "u", "ps"))
```
more efficient, readable, and faster.

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -342,7 +343,10 @@ class Loops(IRNode):
     ranges: List[Expr]
 
     def get_unbacked_symbol_uses(self) -> Set[sympy.Symbol]:
-        return set().union(*(free_unbacked_symbols(e) for e in self.ranges))
+        index = self._index(self.ranges)
+        return set().union(
+            *(free_unbacked_symbols(e) for e in self.ranges)
+        ) | extract_free_unbacked_symbols(self.inner_fn, index)
```
Comment (Skylion007) [inline torch/_inductor/ir.py:349]:
```suggestion
        return set().union(
            *(free_unbacked_symbols(e) for e in self.ranges),
            extract_free_unbacked_symbols(self.inner_fn, index))
```
works I think?

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -1595,12 +1608,18 @@ def index_length(self):
     def inner_fn_str(self):
         index = self._index(self.ranges)
         rindex = self._index(self.scan_ranges, "r")
+        idx = self.reindex(index, rindex)
         return V.KernelFormatterHandler.ir_to_string(
             self.inner_fn,
-            index,
-            rindex,
+            idx,
```
Comment (ezyang) [inline torch/_inductor/ir.py:1614]: @peterbell10 I think I fixed a bug in printing Scans, PTAL.

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 5]
```diff
@@ -587,6 +586,21 @@ def metadata_fn(*primals):
     )
 
 
+class PropagateUnbackedSymInts(torch.fx.Interpreter):
+    def run_node(self, n: torch.fx.Node):
+        import sympy
+
+        result = super().run_node(n)
+        # TODO: handle Tensor returns
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:594]: for my understanding: I would naively thought that:

(1) for an fx node that produces a tensor `x` with unbacked symints, there will be a later fx node that grabs that unbacked symint directly before anyone uses it (`aten.sym_size(x, dim)`)

(2) therefore we should be guaranteed to do the replacement / equality check later, before the symint is used anywhere

Is (1) not guaranteed?

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 6]
```diff
@@ -601,7 +615,9 @@ def functional_call(*args, **kwargs):
                         "ignore", "Anomaly Detection has been enabled."
                     )
                     with torch.autograd.detect_anomaly(check_nan=False):
-                        out = Interpreter(mod).run(*args[params_len:], **kwargs)
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:604]: Also just a thought - we do this propagation in a few places (e.g. inductor re-runs fake prop), so I guess there's a footgun here than anyone might forget this when writing an retracing pass somewhere, and use the default `fx.Interpreter`.

Do you think that foot gun warrants adding this logic to the default `fx.Interpreter`? (I guess we only really do this once in AOTAutograd and once inductor, so maybe this will come up rarely)

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 7]
```diff
@@ -93,7 +93,7 @@ def add_assertions(val):
             call_backs: List[Callable] = []
             messages: List[str] = []
             if isinstance(val, (torch.SymInt, torch.SymFloat, torch.SymBool)):
-                symbol = val.node._expr
```
Comment (bdhirsh) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:96]: how did this blow up / are there any tests this fixes?

File: torch/_functorch/_aot_autograd/traced_function_transforms.py (hunk 6 above)
Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:604]: I thought about this, and my problem is that... it's kind of PT2 specific so Interpreter feels like the wrong place to put it.

File: torch/_functorch/_aot_autograd/traced_function_transforms.py (hunk 5 above)
Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:594]: It is not guaranteed! I actually had homework to come up with a test case that triggers this. Here it is:

```
    @torch._dynamo.config.patch(
        capture_dynamic_output_shape_ops=True
    )
    def test_nonzero_no_realloc(self, device):
        @torch.compile(fullgraph=True)
        def f(x, y):
            z = x.nonzero()
            return torch.split(z, [y.size(0)])

        f(torch.tensor([1, 0, 1, 1, 0, 1, 0]), torch.randn(4))
```

This gives

```
[2024-01-25 08:34:32,887] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_env
[2024-01-25 08:34:32,895] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u0 [-9223372036854775808, 9223372036854775807] at test/inductor/test_torchinductor_dynamic_shapes.py:320 in f (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:32,994] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement u0 = 4
[2024-01-25 08:34:32,995] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] runtime_assert Eq(u0, 4) [guard added] at test/inductor/test_torchinductor_dynamic_shapes.py:321 in f (_decomp/decompositions.py:1229 in split_with_sizes)
[2024-01-25 08:34:32,995] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement u0 = 4
[2024-01-25 08:34:33,013] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u1 [-9223372036854775808, 9223372036854775807] (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:33,309] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u2 [-9223372036854775808, 9223372036854775807] (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:33,331] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] produce_guards
```

The u2 is an unrelated inductor thing. The u1 is from AOTAutograd fake tensor prop; the size of the nonzero tensor never shows up anywhere so we fail to unify it with u0.

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py (hunk 7 above)
Comment (ezyang) [reply torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:96]: The export tests see the old u2 rather than the post replacement u0 when they access this directly. So... yes :)

File: torch/_functorch/_aot_autograd/traced_function_transforms.py (hunk 5 above)
Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:594]: Ah right. In this case though: we never actually **use** the unbacked symint from the nonzero anywhere. I agree that unifying even in this case would be better than not. I guess I'm curious whether this can actually lead to any problems in practice.

There's now logic in inductor to grab every unbacked symint in the graph (even if they are unused) and do stuff to it like track dependencies, so probably better to be safe than sorry, and replace dup'd symints in all cases.

Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:594]: It is true that in this test case, it doesn't actually cause a problem. And I cannot think of an easy way to construct a problematic case. So let's just call this an invariant to be enforced, perhaps for later use.

File: torch/_inductor/dependencies.py [hunk 8]
```diff
@@ -442,3 +450,54 @@ def extract_input_node_reduction_ranges(
...
+        return (None,) * num_values if num_values > 1 else None
+
+
+def _typecheck_FreeUnbackedSymbolsOpsHandler(
+    h: FreeUnbackedSymbolsOpsHandler,
+) -> OpsHandler[None]:
+    return h
+
+
+def extract_free_unbacked_symbols(fn: Callable[..., Any], index, rindex=None):
```
Comment (bdhirsh) [inline torch/_inductor/dependencies.py:492]: I noticed that when I check out your PR locally and put a breakpoint here, it doesn't get by the new test. Probably because.. the test uses `index_select`, which doesn't map to a `Loops` or `Reduction` in inductor's IR?

Can you add a test to exercise this code path?

Comment (bdhirsh) [reply torch/_inductor/dependencies.py:492]: I tried adding `out.sum()` to the existing test and still don't hit it 🤔

**Review by bdhirsh — APPROVED** [review]
I didn't understand the Scan fix (leaving to Peter), but I followed everything else. The new docs helped a ton, thanks!
